src/extent.c

```diff
@@ -1057,11 +1057,19 @@ extent_destroy_wrapper(tsdn_t *tsdn, pac_t *pac, ehooks_t *ehooks,
     edata_t *edata) {
 	assert(edata_base_get(edata) != NULL);
 	assert(edata_size_get(edata) != 0);
+	assert(edata_state_get(edata) == extent_state_retained);
 	witness_assert_depth_to_rank(tsdn_witness_tsdp_get(tsdn),
 	    WITNESS_RANK_CORE, 0);
 
-	/* Deregister first to avoid a race with other allocating threads. */
-	extent_deregister(tsdn, pac, edata);
+	/*
+	 * Deregister to avoid a race with other allocating threads. If guarded,
+	 * then unguard without re-registering in the emap.
+	 */
+	if (edata_guarded_get(edata)) {
+		unguard_pages(tsdn, ehooks, edata, pac->emap, false);
+	} else {
+		extent_deregister(tsdn, pac, edata);
+	}
```

Comment:
If we always do deregister on this path, can we avoid the additional flag for the unguard_pages? It might seem to do redundant operations that way, but since the destroy very rarely happens (if at all), it should be a net win (both simpler API, and one less branch).

Reply:
We can't do it that easily, because `unguard_pages` does `emap_register_boundary`, which has `assert(edata_state_get(edata) == extent_state_active);`. I would prefer to keep the assertion to serve its purpose and catch bugs, because `emap_register_boundary` is used in two more places.

Reply:
I see. In that case, let's do `emap_update_edata_state(tsdn, pac->emap, edata, extent_state_active);` before calling unguard? That way also feels safer since `active` is an acquired state (meaning other threads won't try accessing it), and it can remain only unguard active extents. Then we can still contain the logic to the destroy side instead of tweaking the unguard side.

Reply:
Good point regarding gdump. I'll take a look again tomorrow; a bit out of energy now.

Talking about gdump, don't we have a double `deregsiter`, i.e. double gdump subtraction along the evict => destroy path?

Reply:
We have double `deregister`, but in `extent_gdump_sub` there's a check that subtracts only if the extent is `active`. Tricky.

Reply:
Indeed. Thanks for the discussion -- let's continue and see if this will work:

Context: the deregister happens under ecache_evict makes sense, because we are holding the ecache lock. A general rule we follow is, an edata should either be covered by a state lock (i.e. mutex for the ecache it's part of), or it's in a acquired state.  See this [sanity check function for details](https://github.com/jemalloc/jemalloc/blob/4d56aaeca5883ae5f4b5550c528503fb51fdf479/include/jemalloc/internal/emap.h#L169). Therefore, the deregister before releasing the lock in evict is doing the right thing.
1) let's get rid of the dup deregister in `extent_destroy_wrapper`? We can use the above assert to verify it's acquired in place of the deregister. To be safe, I'd say add a separate commit to make it clear it's intentional.
2) then maybe a new `unguard_pages_pre_destroy()`, which simply calls the `static unguard_pages_impl()` with the `reg_emap == false` (the flag should cover both the deregister and register), while the original `unguard_pages()` calls it with `reg_emap==true`. This way we makes sure only the destroy path gets the special treatment of bypassing emap re-register.

Reply:
Sounds good. This is one of the approaches I've considered, let's implement it!